synapse/handlers/federation.py

```diff
@@ -1035,6 +1043,217 @@ async def get_state_ids_for_pdu(self, room_id: str, event_id: str) -> List[str]:
         else:
             return []
 
+    async def get_backfill_events(
+        self, origin: str, room_id: str, event_id_list: list, limit: int
+    ) -> List[EventBase]:
```

Comment:
This is pretty much functionally equivalent to the `store.get_backfill_events`. I've commented on the new pieces.

Reply:
From a [discussion today](https://docs.google.com/document/d/1KCEmpnGr4J-I8EeaVQ8QJZKBDu53ViI7V62y5BzfXr0/edit#bookmark=id.1hbt9acs963h) with @erikjohnston, we were able to come up with an alternative.

The key breakthrough was discussing all the ways we can find connected insertion events as we're navigating up the DAG and backfilling. We went with option 3 which also gives us a nice simplification:

 1. Find by insertion event `prev_events` (this is what we were doing before)
     - This one seemed nice because the insertion events point by `prev_event` directly off the event we branch from.
     - The problem is deciding whether we should navigate up "live" DAG vs the historical branch first, see https://github.com/matrix-org/synapse/pull/11114#discussion_r731363046. The way I solved this was going back to the federating homeserver and asking whether they have a given insertion event locally.
 1. Find connected insertion events by depth
    - This one seems like it would work but feels like a pretty loose connection (might not actually be) where there are forks in the DAG and the `depth` is the same.
 1. **Find connected insertion events by the marker event**
    - This made the most sense since we already backfill the insertion event
      when a marker event is processed (see [`handle_marker_event`](https://github.com/matrix-org/synapse/blob/6250b95efe88385bb3ec2842d5eb76f42ef762ef/synapse/handlers/federation_event.py#L1098-L1172)).
    - Gets rid of the extra `connected_insertion_event_query` lookup in backfill because we
      know it's already backfilled from the marker processing.
    - And gets rid of the extra federated insertion event lookup we added in this to PR
      to ask whether the homeserver requesting backfill already has the
      insertion event (deciding whether we fork  to the history branch before we go down the "live" DAG, see option 1 in this list for more info)
      

[Mermaid live editor playground link](https://mermaid-js.github.io/mermaid-live-editor/edit/#eyJjb2RlIjoiZmxvd2NoYXJ0IEJUXG4gICAgc3ViZ3JhcGggbGl2ZVxuICAgICAgICBtYXJrZXIxPlwibWFya2VyXCJdIC0tLS0-IEIgLS0tLS0tLS0tLS0tLS0tLS0-IEFcbiAgICBlbmRcbiAgICBcbiAgICBzdWJncmFwaCBiYXRjaDBcbiAgICAgICAgYmF0Y2gwLWJhdGNoW1tcImJhdGNoXCJdXSAtLT4gYmF0Y2gwLTIoKFwiMlwiKSkgLS0-IGJhdGNoMC0xKCgxKSkgLS0-IGJhdGNoMC0wKCgwKSkgLS0-IGJhdGNoMC1pbnNlcnRpb25bL2luc2VydGlvblxcXVxuICAgIGVuZFxuXG4gICAgc3ViZ3JhcGggYmF0Y2gxXG4gICAgICAgIGJhdGNoMS1iYXRjaFtbXCJiYXRjaFwiXV0gLS0-IGJhdGNoMS0yKChcIjJcIikpIC0tPiBiYXRjaDEtMSgoMSkpIC0tPiBiYXRjaDEtMCgoMCkpIC0tPiBiYXRjaDEtaW5zZXJ0aW9uWy9pbnNlcnRpb25cXF1cbiAgICBlbmRcbiAgICBcbiAgICBzdWJncmFwaCBiYXRjaDJcbiAgICAgICAgYmF0Y2gyLWJhdGNoW1tcImJhdGNoXCJdXSAtLT4gYmF0Y2gyLTIoKFwiMlwiKSkgLS0-IGJhdGNoMi0xKCgxKSkgLS0-IGJhdGNoMi0wKCgwKSkgLS0-IGJhdGNoMi1pbnNlcnRpb25bL2luc2VydGlvblxcXVxuICAgIGVuZFxuXG5cbiAgICBiYXRjaDAtaW5zZXJ0aW9uIC0tPiBtZW1iZXJCb2IwKFtcIm0ucm9vbS5tZW1iZXIgKGJvYilcIl0pIC0tPiBtZW1iZXJBbGljZTAoW1wibS5yb29tLm1lbWJlciAoYWxpY2UpXCJdKVxuICAgIGJhdGNoMS1pbnNlcnRpb24gLS0-IG1lbWJlckJvYjEoW1wibS5yb29tLm1lbWJlciAoYm9iKVwiXSkgLS0-IG1lbWJlckFsaWNlMShbXCJtLnJvb20ubWVtYmVyIChhbGljZSlcIl0pXG4gICAgYmF0Y2gyLWluc2VydGlvbiAtLT4gbWVtYmVyQm9iMihbXCJtLnJvb20ubWVtYmVyIChib2IpXCJdKSAtLT4gbWVtYmVyQWxpY2UyKFtcIm0ucm9vbS5tZW1iZXIgKGFsaWNlKVwiXSlcblxuICAgIG1hcmtlcjEgLS4tPiBiYXRjaDAtaW5zZXJ0aW9uQmFzZVxuICAgIGJhdGNoMC1pbnNlcnRpb25CYXNlWy9pbnNlcnRpb25cXF0gLS0tLS0tLS0tLS0tLS0tPiBBXG4gICAgYmF0Y2gwLWJhdGNoIC0uLT4gYmF0Y2gwLWluc2VydGlvbkJhc2VbL2luc2VydGlvblxcXVxuICAgIGJhdGNoMS1iYXRjaCAtLi0-IGJhdGNoMC1pbnNlcnRpb25cbiAgICBiYXRjaDItYmF0Y2ggLS4tPiBiYXRjaDEtaW5zZXJ0aW9uXG4iLCJtZXJtYWlkIjoie1xuICBcInRoZW1lXCI6IFwiZGVmYXVsdFwiXG59IiwidXBkYXRlRWRpdG9yIjpmYWxzZSwiYXV0b1N5bmMiOnRydWUsInVwZGF0ZURpYWdyYW0iOmZhbHNlfQ)
![chrome_2021-11-02_17-44-12](https://user-images.githubusercontent.com/558581/139961953-f9902a1b-495b-4499-baee-984ca1e44d37.png)

Reply:
One problem with finding connected insertion events by relying on marker events to point them out is if marker event was sent somewhere sufficiently long ago and some user just joins now, they have to backfill the marker event before they're able to discover the old history.


The tough scenario:

 1. Imagine a room existing the MSC2716 world for years with lots of "live" messages 
 1. Then the room batch sends a bunch of history at the start of the room, and sends a marker event
 1. Some more years pass in the room with plenty of "live" message activity
 1. A new user joins the room (homeserver that hasn't been in the room)
 1. In order for the new user to be able to discover the batch sent history, they need to come across the marker event randomly in the middle of the "live" history.

---

Maybe this is fine...

A user can only get to those really old historical batch sent messages realistically by permalink or jump to date. Both of those endpoints would return the historical messages even if the marker wasn't discovered.

If someone scrolls back all of that way, they will come across the marker event and it will be fine.